### ISSUE TITLE
[tablets] Decommissioning a rack may fail with std::runtime_error: Unable to find new replica for tablet

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#19475]
As reported by @soyacz:

having tablet's in multi-rack environment, can we actually remove a rack?
E.g. I have such a cluster:
```
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: Datacenter: eu-west
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: ===================
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: Status=Up/Down
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: |/ State=Normal/Leaving/Joining/Moving
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: -- Address    Load    Tokens Owns Host ID                              Rack
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: UN 10.4.0.202 ?       256    ?    20f79832-0544-409a-883a-48ac47294bdd 1a  
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: UN 10.4.0.49  ?       256    ?    76e0506e-db31-43d0-9db2-946516d6cfc5 1a  
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: UN 10.4.1.172 3.86 GB 256    ?    f9208a5b-c947-4e31-ab49-86042c139db8 1a  
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: UN 10.4.3.73  8.16 GB 256    ?    e3448c06-a59f-4ee7-a847-18821422eb2a 1a  
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: UN 10.4.5.155 2.64 GB 256    ?    3809180c-9cf1-4416-9cfc-a3712b524e75 1b  
< t:2024-06-21 16:02:36,727 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.4.5.155>: UN 10.4.8.100 3.36 GB 256    ?    cf78f704-51e6-45f7-ae52-c411b46329de 1c  
```
and tried to decommission node .155(rack 1b):
```
error executing POST request to http://localhost:10000/storage_service/decommission with parameters {}: remote replied with status code 500 Internal Server Error:
std::runtime_error (Decommission failed. See earlier errors (Rolled back: Failed to drain tablets: std::runtime_error (Unable to find new replica for tablet b34853e0-2fe6-11ef-b780-a8d2e5b98a65:1 on 3809180c-9cf1-4416-9cfc-a3712b524e75:0 when draining {3809180c-9cf1-4416-9cfc-a3712b524e75})))
```

**Comment by bhalevy** [comment]
@yarongilor we need to test decommission with asymmetrical rack configuration in dtest - like start with [1,1,2] and rf=3 and make sure we can decommission any node in that dc.

Plus, a more realistic scenario: start with [2,2,2] rf=3, then:
- add node to rack1, decommission node from rack3
- add node to rack2, decommission node from rack3
variation: start with [2,2,2], add 2 nodes to rack1 and rack2, respectively, then decommission the nodes on rack3 in parallel.

**Comment by soyacz** [comment]
Posting test details.
Basically, test started off with 3 nodes in rack 1a, then test added 3 another nodes to different racks: a,b and c.
Then when trying to decommission node from rack 1b it failed with error above.

## Packages

Scylla version: `6.1.0~dev-20240619.7567b87e72cf` with build-id `f0a41eece67ec7a8d64b047e743f0396d9f23463`





Kernel Version: `5.15.0-1063-aws`


## Installation details

Cluster size: 3 nodes (i4i.large)

Scylla Nodes used in this run:
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-9 (18.203.252.160 | 10.4.8.156) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-8 (3.253.127.24 | 10.4.4.33) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-7 (3.251.71.84 | 10.4.0.229) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-6 (54.216.162.126 | 10.4.8.100) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-5 (34.242.23.70 | 10.4.5.155) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-4 (54.229.149.125 | 10.4.1.172) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-3 (54.75.90.72 | 10.4.0.49) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-2 (54.171.108.38 | 10.4.0.202) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-12 (52.213.153.230 | 10.4.11.4) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-11 (34.253.167.115 | 10.4.5.191) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-10 (34.254.243.192 | 10.4.0.165) (shards: 2)
- longevity-5gb-1h-GrowShrinkClusterN-db-node-6ec810eb-1 (54.247.218.10 | 10.4.3.73) (shards: 2)


OS / Image: `ami-038debe0a5d79391a` (aws: undefined_region)

Test: `longevity-5gb-1h-GrowShrinkClusterNemesis-aws-test`
Test id: `6ec810eb-a790-420a-9e21-b4c7499a8812`
Test name: `scylla-staging/lukasz/longevity-5gb-1h-GrowShrinkClusterNemesis-aws-test`
Test config file(s):

- [longevity-5gb-1h-nemesis.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/1b5e6723193e1a76bf8f44bc6a301f68e1ea86f2/configurations/nemesis/longevity-5gb-1h-nemesis.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor 6ec810eb-a790-420a-9e21-b4c7499a8812`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=6ec810eb-a790-420a-9e21-b4c7499a8812)
- Show all stored logs command: `$ hydra investigate show-logs 6ec810eb-a790-420a-9e21-b4c7499a8812`


## Logs:
- **db-cluster-6ec810eb.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/db-cluster-6ec810eb.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/db-cluster-6ec810eb.tar.gz)
- **sct-runner-events-6ec810eb.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/sct-runner-events-6ec810eb.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/sct-runner-events-6ec810eb.tar.gz)
- **sct-6ec810eb.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/sct-6ec810eb.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/sct-6ec810eb.log.tar.gz)
- **loader-set-6ec810eb.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/loader-set-6ec810eb.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/loader-set-6ec810eb.tar.gz)
- **monitor-set-6ec810eb.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/monitor-set-6ec810eb.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/monitor-set-6ec810eb.tar.gz)
- **parallel-timelines-report-6ec810eb.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/parallel-timelines-report-6ec810eb.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/6ec810eb-a790-420a-9e21-b4c7499a8812/20240621_170615/parallel-timelines-report-6ec810eb.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-staging/job/lukasz/job/longevity-5gb-1h-GrowShrinkClusterNemesis-aws-test/19/)
[Argus](https://argus.scylladb.com/test/a2af091e-3cbd-4068-b825-35d511f15d17/runs?additionalRuns[]=6ec810eb-a790-420a-9e21-b4c7499a8812)
</details>

**Comment by bhalevy** [comment]
> Basically, test started off with 3 nodes in rack 1a, then test added 3 another nodes to different racks: a,b and c.

@soyacz but your report in https://github.com/scylladb/scylladb/issues/19475#issue-2371847519 shows that all new nodes were added to rack 1a, so maybe there is also an issue in the test.

**Comment by bhalevy** [comment]
cc @xemul @tgrabiec

**Comment by soyacz** [comment]
> > Basically, test started off with 3 nodes in rack 1a, then test added 3 another nodes to different racks: a,b and c.
> 
> @soyacz but your report in [#19475 (comment)](https://github.com/scylladb/scylladb/issues/19475#issue-2371847519) shows that all new nodes were added to rack 1a, so maybe there is also an issue in the test.

no, test started with 3 nodes in 1a and then added to different racks. 
And yes, issue was with the test, as it should have started with 1a,1b,1c racks - anyway I asked question if this scenario shouldn't be supported.

**Comment by juliayakovlev** [comment]
Similar issue receive with single rack environment

Nodetool decommission failed:
```
Jun 23 06:55:03.627291 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
Jun 23 06:55:54.545232 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] raft_topology - coordinator fiber found a node to work on id=762258fd-11de-44f0-8c3b-f29bb81a0b44 state=normal
Jun 23 06:55:54.545278 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] raft_topology - updating topology state: start decommission
Jun 23 06:55:54.569008 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] raft_topology - entered `tablet draining` transition state
Jun 23 06:55:54.569029 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Examining DC eu-west (shuffle=false, balancing=true)
Jun 23 06:55:54.569044 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Will drain node 762258fd-11de-44f0-8c3b-f29bb81a0b44 (being_decommissioned) from DC eu-west

Jun 23 06:55:54.601248 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Node c95115fd-64cb-4de0-90aa-b29f71371e5e: rack=1c avg_load=17.966666666666665, tablets=539, shards=30, state=normal
Jun 23 06:55:54.601849 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - target node: 57d2df2f-e8c1-474b-a7c1-1396fcb7947e, avg_load: 17.933333333333334, max: 17.966666666666665
Jun 23 06:55:54.602256 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] raft_topology - tablets draining failed with std::runtime_error (Unable to find new replica for tablet 4eb84450-3084-11ef-993e-55b6485231a6:30 on 762258fd-11de-44f0-8c3b-f29bb81a0b44:21 when draining {762258fd-11de-44f0-8c3b-f29bb81a0b44}). Aborting the topology operation
Jun 23 06:55:54.602281 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] raft_topology - barrier for tablet 402e58a0-3054-11ef-aa28-5d41776339ca:44 failed: seastar::broken_promise (broken promise)
Jun 23 06:55:54.602290 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] raft_topology - barrier for tablet 4ed93aa1-3054-11ef-9d97-c4fa041d5a43:62 failed: seastar::broken_promise (broken promise)
```

## Packages

Scylla version: `6.1.0~dev-20240619.7567b87e72cf` with build-id `f0a41eece67ec7a8d64b047e743f0396d9f23463`

Kernel Version: `5.15.0-1063-aws`


## Issue description

- [ ] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

*Describe your issue in detail and steps it took to produce it.*

## Impact

*Describe the impact this issue causes to the user.*

## How frequently does it reproduce?

*Describe the frequency with how this issue can be reproduced.*

## Installation details

Cluster size: 5 nodes (i4i.8xlarge)

Scylla Nodes used in this run:
- longevity-mv-si-4d-master-db-node-7f60cd4d-9 (34.245.149.16 | 10.4.9.109) (shards: 30)
- longevity-mv-si-4d-master-db-node-7f60cd4d-8 (54.216.161.148 | 10.4.10.92) (shards: 30)
- longevity-mv-si-4d-master-db-node-7f60cd4d-7 (34.244.237.169 | 10.4.9.227) (shards: 30)
- longevity-mv-si-4d-master-db-node-7f60cd4d-6 (18.202.79.191 | 10.4.8.209) (shards: 30)
- longevity-mv-si-4d-master-db-node-7f60cd4d-5 (18.201.177.3 | 10.4.10.26) (shards: 30)
- longevity-mv-si-4d-master-db-node-7f60cd4d-4 (18.202.248.32 | 10.4.8.87) (shards: 30)
- longevity-mv-si-4d-master-db-node-7f60cd4d-3 (54.220.196.232 | 10.4.11.119) (shards: 30)
- longevity-mv-si-4d-master-db-node-7f60cd4d-2 (34.246.121.137 | 10.4.9.216) (shards: 30)
- longevity-mv-si-4d-master-db-node-7f60cd4d-1 (34.248.229.42 | 10.4.10.109) (shards: 30)


OS / Image: `ami-038debe0a5d79391a` (aws: undefined_region)

Test: `longevity-mv-si-4days-test`
Test id: `7f60cd4d-dffc-405a-b337-753f1af79053`
Test name: `scylla-master/tier1/longevity-mv-si-4days-test`
Test config file(s):

- [longevity-mv-si-4days.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/9b2a1cd9d33e7d352721cc77030ad3fe6cfa3b8b/test-cases/longevity/longevity-mv-si-4days.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor 7f60cd4d-dffc-405a-b337-753f1af79053`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=7f60cd4d-dffc-405a-b337-753f1af79053)
- Show all stored logs command: `$ hydra investigate show-logs 7f60cd4d-dffc-405a-b337-753f1af79053`


## Logs:
- **core.scylla-longevity-mv-si-4d-master-db-node-7f60cd4d-5-2024-06-22_15-26-56.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.b30872317c7d48c49fb6bba0d8822419.22750.1719068539000000/core.scylla.112.b30872317c7d48c49fb6bba0d8822419.22750.1719068539000000.gz](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.b30872317c7d48c49fb6bba0d8822419.22750.1719068539000000/core.scylla.112.b30872317c7d48c49fb6bba0d8822419.22750.1719068539000000.gz)
- **core.scylla-longevity-mv-si-4d-master-db-node-7f60cd4d-1-2024-06-22_15-27-17.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.c1e1e1333e25415eafe6372ddf312d10.25180.1719068547000000/core.scylla.112.c1e1e1333e25415eafe6372ddf312d10.25180.1719068547000000.gz](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.c1e1e1333e25415eafe6372ddf312d10.25180.1719068547000000/core.scylla.112.c1e1e1333e25415eafe6372ddf312d10.25180.1719068547000000.gz)
- **core.scylla-longevity-mv-si-4d-master-db-node-7f60cd4d-3-2024-06-22_15-28-25.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.83213b980af841b5830b7f9ed27be49f.23901.1719068539000000/core.scylla.112.83213b980af841b5830b7f9ed27be49f.23901.1719068539000000.gz](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.83213b980af841b5830b7f9ed27be49f.23901.1719068539000000/core.scylla.112.83213b980af841b5830b7f9ed27be49f.23901.1719068539000000.gz)
- **core.scylla-longevity-mv-si-4d-master-db-node-7f60cd4d-4-2024-06-22_15-28-37.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.c011e1a9a9ba40108b02caa80892b900.22855.1719068559000000/core.scylla.112.c011e1a9a9ba40108b02caa80892b900.22855.1719068559000000.gz](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.c011e1a9a9ba40108b02caa80892b900.22855.1719068559000000/core.scylla.112.c011e1a9a9ba40108b02caa80892b900.22855.1719068559000000.gz)
- **core.scylla-longevity-mv-si-4d-master-db-node-7f60cd4d-2-2024-06-22_15-31-01.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.f364312289fb4baaa9783344d0a549cc.24142.1719068539000000/core.scylla.112.f364312289fb4baaa9783344d0a549cc.24142.1719068539000000.gz](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.112.f364312289fb4baaa9783344d0a549cc.24142.1719068539000000/core.scylla.112.f364312289fb4baaa9783344d0a549cc.24142.1719068539000000.gz)
- **db-cluster-7f60cd4d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/db-cluster-7f60cd4d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/db-cluster-7f60cd4d.tar.gz)
- **sct-runner-events-7f60cd4d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/sct-runner-events-7f60cd4d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/sct-runner-events-7f60cd4d.tar.gz)
- **sct-7f60cd4d.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/sct-7f60cd4d.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/sct-7f60cd4d.log.tar.gz)
- **loader-set-7f60cd4d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/loader-set-7f60cd4d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/loader-set-7f60cd4d.tar.gz)
- **monitor-set-7f60cd4d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/monitor-set-7f60cd4d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7f60cd4d-dffc-405a-b337-753f1af79053/20240623_074122/monitor-set-7f60cd4d.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/tier1/job/longevity-mv-si-4days-test/20/)
[Argus](https://argus.scylladb.com/test/48122dfe-148b-4a99-a43c-1a71c6869973/runs?additionalRuns[]=7f60cd4d-dffc-405a-b337-753f1af79053)
</details>

**Comment by bhalevy** [comment]
> Similar issue receive with single rack environment

@juliayakovlev how many nodes are in the datacenter?
What is the replication factor?

If rf == num_nodes then decommission is expected to fail with tablets since there aren't enough remaining nodes to satisfy the replication factor.

**Comment by juliayakovlev** [comment]
> > Similar issue receive with single rack environment
> 
> @juliayakovlev how many nodes are in the datacenter? What is the replication factor?
> 
> If rf == num_nodes then decommission is expected to fail with tablets since there aren't enough remaining nodes to satisfy the replication factor.

there are 5 nodes in cluster (one DC). Keyspaces RF=3

**Comment by bhalevy** [comment]
```
Jun 23 06:55:54.601084 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Will drain node 762258fd-11de-44f0-8c3b-f29bb81a0b44 (being_decommissioned) from DC eu-west
Jun 23 06:55:54.601235 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Node 762258fd-11de-44f0-8c3b-f29bb81a0b44: rack=1c avg_load=16.7, tablets=501, shards=30, state=being_decommissioned
Jun 23 06:55:54.601241 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Node 00e2305f-e0b9-4896-81b7-9f2e1e529dc2: rack=1c avg_load=17.966666666666665, tablets=539, shards=30, state=normal
Jun 23 06:55:54.601243 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Node d105c73b-8556-4ff1-8f4c-102b9eec536a: rack=1c avg_load=17.966666666666665, tablets=539, shards=30, state=normal
Jun 23 06:55:54.601246 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Node 57d2df2f-e8c1-474b-a7c1-1396fcb7947e: rack=1c avg_load=17.933333333333334, tablets=538, shards=30, state=normal
Jun 23 06:55:54.601248 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - Node c95115fd-64cb-4de0-90aa-b29f71371e5e: rack=1c avg_load=17.966666666666665, tablets=539, shards=30, state=normal
Jun 23 06:55:54.601849 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] load_balancer - target node: 57d2df2f-e8c1-474b-a7c1-1396fcb7947e, avg_load: 17.933333333333334, max: 17.966666666666665
Jun 23 06:55:54.602256 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[29606]:  [shard  0: gms] raft_topology - tablets draining failed with std::runtime_error (Unable to find new replica for tablet 4eb84450-3084-11ef-993e-55b6485231a6:30 on 762258fd-11de-44f0-8c3b-f29bb81a0b44:21 when draining {762258fd-11de-44f0-8c3b-f29bb81a0b44}). Aborting the topology operation
```

**Comment by bhalevy** [comment]
Hmm, there was a previous failed decommission event that hit a different tablet in this table.
I wonder if they could be related...
```
Jun 23 00:24:47.039224 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[7008]:  [shard  2:strm] api - decommission
Jun 23 00:24:47.039891 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[7008]:  [shard  0:strm] raft_topology - request decommission for: d105c73b-8556-4ff1-8f4c-102b9eec536a
Jun 23 00:24:47.442089 longevity-mv-si-4d-master-db-node-7f60cd4d-7 scylla[7008]:  [shard  0:strm] raft_topology - Decommission failed. See earlier errors (Rolled back: Failed to drain tablets: std::runtime_error (Unable to find new replica for tablet 4eb84450-3084-11ef-993e-55b6485231a6:17 on d105c73b-8556-4ff1-8f4c-102b9eec536a:6 when draining {d105c73b-8556-4ff1-8f4c-102b9eec536a}))
```

**Comment by bhalevy** [comment]
@juliayakovlev in https://github.com/scylladb/scylladb/issues/19475#issuecomment-2188843371, although the symptom is similar, it seems like it has a different root cause.
Can you please a separate issue for that.
Worst case, w can close it as a dup, but I doubt it is.

**Comment by xemul** [comment]
@bhalevy , it looks like decommissioning a rack is impossible by design. There's a [test](https://github.com/scylladb/scylladb/blob/263668bc85b2e1485c4bf11efb372252e4b0282a/test/boost/tablets_test.cc#L1420) for that:

```c++
SEASTAR_THREAD_TEST_CASE(test_decommission_rack_load_failure) {
    // Verifies that load balancer moves tablets out of the decommissioned node.
    // The scenario is such that it is impossible to distribute replicas without violating rack uniqueness.
    ...
            tm.update_topology(host1, racks[0], std::nullopt, shard_count);
            tm.update_topology(host2, racks[0], std::nullopt, shard_count);
            tm.update_topology(host3, racks[0], std::nullopt, shard_count);
            tm.update_topology(host4, racks[1], node::state::being_decommissioned,
                               shard_count);
            ...
            BOOST_REQUIRE_THROW(rebalance_tablets(e.get_tablet_allocator().local(), stm), std::runtime_error);
```

So 3 nodes in rack-a and one node in rack-b being decommissioned -> expected failure

@tgrabiec , please comment

**Comment by tgrabiec** [comment]
> @bhalevy , it looks like decommissioning a rack is impossible by design. There's a [test](https://github.com/scylladb/scylladb/blob/263668bc85b2e1485c4bf11efb372252e4b0282a/test/boost/tablets_test.cc#L1420) for that:
> 
> ```c++
> SEASTAR_THREAD_TEST_CASE(test_decommission_rack_load_failure) {
>     // Verifies that load balancer moves tablets out of the decommissioned node.
>     // The scenario is such that it is impossible to distribute replicas without violating rack uniqueness.
>     ...
>             tm.update_topology(host1, racks[0], std::nullopt, shard_count);
>             tm.update_topology(host2, racks[0], std::nullopt, shard_count);
>             tm.update_topology(host3, racks[0], std::nullopt, shard_count);
>             tm.update_topology(host4, racks[1], node::state::being_decommissioned,
>                                shard_count);
>             ...
>             BOOST_REQUIRE_THROW(rebalance_tablets(e.get_tablet_allocator().local(), stm), std::runtime_error);
> ```
> 
> So 3 nodes in rack-a and one node in rack-b being decommissioned -> expected failure
> 
> @tgrabiec , please comment

Yes, reducing rack-rf during decommission is prohibited, like reducing (effective) RF. 

If we want to support decommissioning a rack, then we probably need a separate topology operation for that. Just decommissioning nodes is not a good idea because as you remove nodes, and load balancer tries to maintain rack-rf, the few nodes in the shrinking rack would get overloaded with replicas before it becomes not possible to decommission. So better would be to mark the rack as being decommissioned so that load balancer ignores it when calculating rack-rf.

**Comment by tgrabiec** [comment]
Replacing a rack should be fine, if you add nodes in the new rack in tandem with decommissioning in the old rack.

**Comment by fruch** [comment]
@soyacz test is probably with multiple racks, but what's @juliayakovlev reported here is not.

anyhow if we have specific limitation on racks, it should documented / mitigated in SCT.
we don't have lots of case we enable multiple racks, but we do have few.

**Comment by soyacz** [comment]
> anyhow if we have specific limitation on racks, it should documented / mitigated in SCT. we don't have lots of case we enable multiple racks, but we do have few.

basically, we don't drop racks in SCT, this case was only one due to bug in SCT (on my branch that I work on).

**Comment by soyacz** [comment]
> > > anyhow if we have specific limitation on racks, it should documented / mitigated in SCT. we don't have lots of case we enable multiple racks, but we do have few.
> > 
> > 
> > basically, we don't drop racks in SCT, this case was only one due to bug in SCT (on my branch that I work on).
> 
> IIRC the test adds 3 nodes then decommissions 3 nodes. Are the decommissioned nodes selected at random? If so, the test could end up decommissioning a rack inadvertently. For example: you start with 3 nodes, 1 in each rack. Then add 3 nodes, to get to 2 nodes in each rack. decommissioning any 3 random nodes may not pick 1 node from each rack, but 2 nodes from 1 of the racks.

The current state is, that test aims to decommission from each rack evenly in basic scenario (random node from each rack). Current test scenarios when use 3 racks, always grow/shrink by multiple of 3, so we shouldn't get into state of decommissioning rack.

**Comment by xemul** [comment]
> ... so we shouldn't get into state of decommissioning rack.

We shouldn't but the initial Benny's description clearly states we did ust that

> E.g. I have such a cluster:
> p:DEBUG > <10.4.5.155>: UN 10.4.0.202 ?       > 256    ?    20f79832-0544-409a-883a-48ac47294bdd 1a  
> p:DEBUG > <10.4.5.155>: UN 10.4.0.49  ?       > 256    ?    76e0506e-db31-43d0-9db2-946516d6cfc5 1a  
> p:DEBUG > <10.4.5.155>: UN 10.4.1.172 > 3.86 GB 256    ?    f9208a5b-c947-4e31-ab49-86042c139db8 1a  
> p:DEBUG > <10.4.5.155>: UN 10.4.3.73  > 8.16 GB 256    ?    e3448c06-a59f-4ee7-a847-18821422eb2a 1a  
> p:DEBUG > <10.4.5.155>: UN 10.4.5.155 2.64 GB 256    ?    3809180c-9cf1-4416-9cfc-a3712b524e75 1b  
> p:DEBUG > <10.4.5.155>: UN 10.4.8.100 3.36 GB 256    ?    cf78f704-51e6-45f7-ae52-c411b46329de 1c  
> 
> and tried to decommission node .155(rack 1b):

**Comment by temichus** [comment]
last time is seen in SCT tier1 2024-06-22

**Comment by mykaul** [comment]
This is anyway an open issue - currently we cannot decommission a rack with tablets. Unsure we need more than 1 issue for it.

**Comment by temichus** [comment]
> > last time is seen in SCT tier1 2024-06-22
> 
> can you point to it ? the original report is not from tier1 case.
> 
> and @juliayakovlev opened a new one at #19504

https://argus.scylladb.com/test/48122dfe-148b-4a99-a43c-1a71c6869973/runs?additionalRuns[]=7f60cd4d-dffc-405a-b337-753f1af79053

**Comment by fruch** [comment]
> > > last time is seen in SCT tier1 2024-06-22
> > 
> > can you point to it ? the original report is not from tier1 case.
> > 
> > and @juliayakovlev opened a new one at #19504
> 
> https://argus.scylladb.com/test/48122dfe-148b-4a99-a43c-1a71c6869973/runs?additionalRuns[]=7f60cd4d-dffc-405a-b337-753f1af79053

So yes, that's what Julia opened #19504 for.

This one is talking about clearing a rack

**Comment by yarongilor** [comment]
> 1. start with [2,2,2] rf=3 and decommission both nodes in rack3 (we have 4 nodes left, this should work]
> 2. start with [1,1,2] rf=3, then:
> 
> * add node to rack1, decommission node from rack3
> * add node to rack2, decommission node from rack3
> 
> 3. variation: start with [1,1,2], add 2 nodes to rack1 and rack2, respectively, then decommission the nodes on rack3 in parallel.

The above cases could be added to `topology_test.py`, i guess.
It is also possible to add a "decommission rack" test and see verify it is failing as expected.

**Comment by mykaul** [comment]
@yarongilor - what's the status of this issue?

**Comment by yarongilor** [comment]
> @yarongilor - what's the status of this issue?

@mykaul , it is open, yet, has a confusing title.

@bhalevy , i would consider renaming this issue to: "A full rack decommission is not supported"
Once this issue is fixed, the test of `test_decommission_with_multi_rack` should fail and should be revised with the new solution, like suggested by @tgrabiec above.
It could be covered in a revised SCT nemesis as well.

**Comment by bhalevy** [comment]
Hmm, I'm trying to reproduce this with a unit test, but it seems decommissioning the last node in a rack is already possible if there are enough remaining racks to satisfy the replication factor.

**Comment by bhalevy** [comment]
Apparently the problem happens when tablet draining (as part of decommission) cannot find a target node due to imbalance.  When I start a single-dc cluster with 3 racks, 2 nodes in each rack, rf=3, and decommission the nodes in the last rack, the first decommission succeeds, but the second one fails to find a target node.

However with 3 racks, 1 node per rack, and rf=2 decommissioning a node in one of the racks succeeds.

**Comment by bhalevy** [comment]
@tgrabiec The case I mentioned above passes with the following change:
```diff
diff --git a/service/tablet_allocator.cc b/service/tablet_allocator.cc
index d0cf1bf4f4..ac13d79e70 100644
--- a/service/tablet_allocator.cc
+++ b/service/tablet_allocator.cc
@@ -1129,7 +1129,7 @@ class load_balancer {
                 auto old_i = i++;
                 if (src_info.rack() != t_info.rack()) {
                     auto new_rack_load = rack_load[t_info.rack()] + 1;
-                    if (new_rack_load > max_rack_load) {
+                    if (new_rack_load > max_rack_load + 1) {
                         viable_targets.erase(old_i);
                     }
                 }
```

This allows decommissioning across racks where that will create a temporary imbalance.
The question is whether it's acceptable.

**Comment by tgrabiec** [comment]
> This allows decommissioning across racks where that will create a temporary imbalance.

What guarantees that it's temporary?

**Comment by bhalevy** [comment]
> > This allows decommissioning across racks where that will create a temporary imbalance.
> 
> What guarantees that it's temporary?

Next round, the maximum is incremented by 1, so one of the other racks would be chosen.

In any case, I'll publish a PR that had a more subtle change that basically restricts this check if min_rack_load < max_rack_load.  This way, if all other racks are loaded equally we allow adding a replica to any if them, otherwise we erase the racks with max_rack_load from the viable_targets set

**Comment by tgrabiec** [comment]
> > > This allows decommissioning across racks where that will create a temporary imbalance.
> > 
> > 
> > What guarantees that it's temporary?
> 
> Next round, the maximum is incremented by 1, so one of the other racks would be chosen.
> 
> In any case, I'll publish a PR that had a more subtle change that basically restricts this check if min_rack_load < max_rack_load. This way, if all other racks are loaded equally we allow adding a replica to any if them, otherwise we erase the racks with max_rack_load from the viable_targets set

But there's no guarantee that the tablet will be moved again, we can stay in rack overload, no?

**Comment by bhalevy** [comment]
> > > > This allows decommissioning across racks where that will create a temporary imbalance.
> > > 
> > > 
> > > What guarantees that it's temporary?
> > 
> > 
> > Next round, the maximum is incremented by 1, so one of the other racks would be chosen.
> > In any case, I'll publish a PR that had a more subtle change that basically restricts this check if min_rack_load < max_rack_load. This way, if all other racks are loaded equally we allow adding a replica to any if them, otherwise we erase the racks with max_rack_load from the viable_targets set
> 
> But there's no guarantee that the tablet will be moved again, we can stay in rack overload, no?

An overload of 1 is better than failing decommission unpredictably

**Comment by bhalevy** [comment]
Based on the discussion in #20673, it was decided to close this issue as unplanned
and enshrine the current behavior.  Test are added for positive and negative scenarios
of rack decommission in https://github.com/scylladb/scylladb/pull/20747
